# Using Perplexity API
- https://docs.perplexity.ai/guides/getting-started

In [ ]:
! pip install keyring openai

In [1]:
import keyring # pip install keyring
from openai import OpenAI # pip install openai

In [2]:
# Set only once. Do not keep this code (with visible API key) in .py, .ipynb etc.
#keyring.set_password('Perplexity_API', 'API_test', '<API-KEY>')

In [2]:
client = OpenAI(api_key=keyring.get_password('Perplexity_API', 'API_test'), base_url="https://api.perplexity.ai")

In [3]:
MODEL = "sonar" # sonar, sonar-pro - https://docs.perplexity.ai/guides/model-cards

In [5]:
user_prompt = "Kā izveidot API atslēgu perplexity.ai rīkā?"

In [6]:
messages = [
    {
        "role": "system",
        "content": (
            "You are an artificial intelligence assistant and you need to "
            "engage in a helpful, detailed, polite conversation with a user."
        ),
    },
    {   
        "role": "user",
        "content": (
            user_prompt            
        ),
    },
]

In [7]:
messages

[{'role': 'system',
  'content': 'You are an artificial intelligence assistant and you need to engage in a helpful, detailed, polite conversation with a user.'},
 {'role': 'user', 'content': 'Kā izveidot API atslēgu perplexity.ai rīkā?'}]

In [8]:
# chat completion without streaming
response = client.chat.completions.create(
    model=MODEL, # sonar, sonar-pro - https://docs.perplexity.ai/guides/model-cards
    messages=messages,
)
print(response)

ChatCompletion(id='60029789-91f7-4cdb-a715-99a8cd21ebae', choices=[Choice(finish_reason='stop', index=0, logprobs=None, message=ChatCompletionMessage(content='Lai izveidotu API atslēgu **perplexity.ai** rīkā, jāveic šādi:\n\n1. **Reģistrācija**:\n   - Vispirms, jāpievienojas **perplexity.ai** platformai. Šo var izdarīt, apmeklējot **API Settings** lapu.\n   - Reģistrācijas procesā jānorāda maksājumu informācija, lai varētu izmantot API. Šis solis neizraisīs nekādu uzreizēju maksājumu; tas tikai saglabā informāciju nākamajām darbībām.\n\n2. **API atslēgas izveidošana**:\n   - Pēc reģistrācijas varat izveidot savu API atslēgu. Šī atslēga ir ilglaicīgs piekļuves tokens, kas ir nepieciešams katrā API pieprasījumā.\n   - API atslēgu jānosūta kā **Bearer** tokens galvenes daļā ar katru API pieprasījumu.\n\n3. **API atslēgas izmantošana**:\n   - Pēc atslēgas izveidošanas varat to izmantot savās programmās, piemēram, Python vai Java, lai veiktu pieprasījumus uz **perplexity.ai** API.\n\nŠādi v

Response result:

In [11]:
print(response.choices[0].message.content)

Lai izveidotu API atslēgu **perplexity.ai** rīkā, jāveic šādi:

1. **Reģistrācija**:
   - Vispirms, jāpievienojas **perplexity.ai** platformai. Šo var izdarīt, apmeklējot **API Settings** lapu.
   - Reģistrācijas procesā jānorāda maksājumu informācija, lai varētu izmantot API. Šis solis neizraisīs nekādu uzreizēju maksājumu; tas tikai saglabā informāciju nākamajām darbībām.

2. **API atslēgas izveidošana**:
   - Pēc reģistrācijas varat izveidot savu API atslēgu. Šī atslēga ir ilglaicīgs piekļuves tokens, kas ir nepieciešams katrā API pieprasījumā.
   - API atslēgu jānosūta kā **Bearer** tokens galvenes daļā ar katru API pieprasījumu.

3. **API atslēgas izmantošana**:
   - Pēc atslēgas izveidošanas varat to izmantot savās programmās, piemēram, Python vai Java, lai veiktu pieprasījumus uz **perplexity.ai** API.

Šādi varat sākt izmantot **perplexity.ai** API savās aplikācijās, izmantojot izveidoto API atslēgu[2][3].


Used tokens for this query:

In [25]:
print(f"""
Prompt tokens:\t\t{response.usage.prompt_tokens}
Completion tokens: \t{response.usage.completion_tokens}
Total tokens:\t\t{response.usage.total_tokens}
""")


Prompt tokens:		41
Completion tokens: 	396
Total tokens:		437



This query (437 tokens) cost me **0.005 EUR**. Detailed information about pricing you can see [here](https://docs.perplexity.ai/guides/pricing).

Streaming

In [7]:
# chat completion with streaming
response_stream = client.chat.completions.create(
    model=MODEL,
    messages=messages,
    stream=True,
)
for response in response_stream:
    print(response)

ChatCompletionChunk(id='e9b2928f-c787-4a94-928f-f3e9e099c770', choices=[Choice(delta=ChoiceDelta(content='L', function_call=None, refusal=None, role='assistant', tool_calls=None), finish_reason=None, index=0, logprobs=None, message={'role': 'assistant', 'content': 'L'})], created=1737709489, model='sonar', object='chat.completion', service_tier=None, system_fingerprint=None, usage=CompletionUsage(completion_tokens=1, prompt_tokens=41, total_tokens=42, completion_tokens_details=None, prompt_tokens_details=None), citations=['https://www.youtube.com/watch?v=LOe2FMuBpT8', 'https://docs.perplexity.ai/guides/getting-started', 'https://www.youtube.com/watch?v=sNH_s9UWVVc', 'https://www.youtube.com/watch?v=46XRqjOjzE0', 'https://community.appinventor.mit.edu/t/perplexity-chatbot-project/102186'])
ChatCompletionChunk(id='e9b2928f-c787-4a94-928f-f3e9e099c770', choices=[Choice(delta=ChoiceDelta(content='ai iz', function_call=None, refusal=None, role='assistant', tool_calls=None), finish_reason=No